### PR TITLE
WIP: DeterministicKeyChain more cleanup

### PR DESCRIPTION
**Pull request by msgilligan**
A series of commits that improve `DeterministicKeychain` (and one that improves a test to better use `HDPath`)

### CODE REVIEW

**Comment by schildbach** [comment]
Maybe rebase, to see what's left?

**Comment by msgilligan** [comment]
> Maybe rebase, to see what's left?

I've rebased/force-pushed and I've submitted the first commit from this branch as PR #2408.

**Comment by schildbach** [comment]
I've merged a couple of those directly, so maybe it's time for another rebase?

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java [hunk 1]
```diff
@@ -804,7 +804,7 @@ static List<DeterministicKeyChain> fromProtobuf(List<Protos.Key> keys, @Nullable
      * key rotation it can happen that there are multiple chains found.
      */
     public static List<DeterministicKeyChain> fromProtobuf(List<Protos.Key> keys, @Nullable KeyCrypter crypter, KeyChainFactory factory) throws UnreadableWalletException {
-        List<DeterministicKeyChain> chains = new LinkedList<>();
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:800]: Given that we generally do paste `final` to every variable (even if used as a value), I wonder if this change is in line with the codebase?

**Comment by msgilligan** [comment]
@schildbach rebased and force-pushed just now.

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java (hunk 1 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:802]: Given that this particular method has 170 lines of code and has many local variables that are mutated in complex ways, I thought adding `final` here was worth doing. The other 6 variables declared before the for loop are all mutated inside the loop.

**Comment by msgilligan** [comment]
Some of this work is in PR #2421. I'll make this a DRAFT until that one is merged.

**Comment by msgilligan** [comment]
Some of this work is in PR #2421. I'll make this a DRAFT until that one is merged.

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java (hunk 1 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:802]: I agree we don't need `final` here.

**Comment by msgilligan** [comment]
After PR #2427 is merged, everything we want from this PR will have been merged.

**Pull request closed without merging**